### PR TITLE
CHE-11954: Add che.registerTaskRunner Theia Plugin API

### PR DESCRIPTION
**Pull request by vinokurig**
fixes https://github.com/eclipse/che/issues/11954

### CODE REVIEW

File: extensions/eclipse-che-theia-plugin-ext/src/browser/che-task-main.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+export class CheTaskMainImpl implements CheTaskMain{
+    private readonly delegate: CheTaskService;
+    private readonly cheTaskClient: CheTaskClient;
+    constructor(container: interfaces.Container, rpc: RPCProtocol) {
+        const proxy: CheTask = rpc.getProxy(PLUGIN_RPC_CONTEXT.CHE_TASK);
+        this.delegate = container.get(CheTaskService);
+        this.cheTaskClient = container.get(CheTaskClient);
+        this.cheTaskClient.onKillEvent(id => {
+            proxy.$killTask(id);
+        });
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply extensions/eclipse-che-theia-plugin-ext/src/browser/che-task-main.ts]: fixed

File: extensions/eclipse-che-theia-plugin-ext/src/browser/che-task-main.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        this.cheTaskClient.setTaskInfoHandler(id => proxy.$getTaskInfo(id));
+        this.cheTaskClient.setRunTaskHandler((config, ctx) => proxy.$runTask(config, ctx));
+    }
+    async $registerTaskRunner(type: string): Promise<void> {
+        return await this.delegate.registerTaskRunner(type);
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply extensions/eclipse-che-theia-plugin-ext/src/browser/che-task-main.ts]: fixed

File: extensions/eclipse-che-theia-plugin-ext/src/browser/che-task-main.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    async $disposeTaskRunner(type: string): Promise<void> {
+        await this.delegate.disposeTaskRunner(type);
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply extensions/eclipse-che-theia-plugin-ext/src/browser/che-task-main.ts]: fixed

File: extensions/eclipse-che-theia-plugin-ext/src/node/che-backend-module.ts [hunk 4]
```diff
@@ -30,4 +37,17 @@ export default new ContainerModule(bind => {
...
     ).inSingletonScope();
+
+    bind(CheTaskService).toDynamicValue(ctx => new CheTaskServiceImpl(ctx.container)).inSingletonScope();
+    bind(ConnectionHandler).toDynamicValue(ctx =>
+        new JsonRpcConnectionHandler<CheTaskClient>(CHE_TASK_SERVICE_PATH, client => {
+                const server: CheTaskService = ctx.container.get(CheTaskService);
+                server.setClient(client);
+                client.onDidCloseConnection(() => {
+                    server.disconnectClient(client);
+                });
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply extensions/eclipse-che-theia-plugin-ext/src/node/che-backend-module.ts]: fixed

File: extensions/eclipse-che-theia-plugin/src/che-proposed.d.ts [hunk 5]
```diff
@@ -335,4 +335,43 @@ declare module '@eclipse-che/plugin' {
         dispose(): PromiseLike<void>;
     }
 
+    export namespace task {
+        export function registerTaskRunner(type: string, runner: TaskRunner): Promise<Disposable>;
+    }
+
+    /** A Task Runner knows how to setConfig and kill a Task of a particular type. */
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply extensions/eclipse-che-theia-plugin/src/che-proposed.d.ts]: fixed

**Comment by vitaliy-guliy** [comment]
Travis failed with
````
src/node/che-task-service.ts(84,11): error TS2416: Property 'getRuntimeInfo' in type 'CheTask' is not assignable to the same property in base type 'Task'.
```

**Comment by vinokurig** [comment]
@vitaliy-guliy that's because the PR depends on https://github.com/theia-ide/theia/pull/3909, when theia 3.19 will be released it will be fixed.

**Comment by evidolob** [comment]
@vinokurig there is some errors during build:
```
src/node/che-task-service.ts(45,13): error TS2322: Type 'CheTask' is not assignable to type 'Task'.
  Types of property 'getRuntimeInfo' are incompatible.
    Type '() => Promise<TaskInfo>' is not assignable to type '() => TaskInfo'.
src/node/che-task-service.ts(84,11): error TS2416: Property 'getRuntimeInfo' in type 'CheTask' is not assignable to the same property in base type 'Task'.
  Type '() => Promise<TaskInfo>' is not assignable to type '() => TaskInfo'.
    Type 'Promise<TaskInfo>' is not assignable to type 'TaskInfo'.
      Property 'taskId' is missing in type 'Promise<TaskInfo>'.
```

**Comment by vinokurig** [comment]
@evidolob It is expected, I've merged some related changes to Theia repo (https://github.com/theia-ide/theia/pull/3909) and since it is not released, the PR build fails.

**Comment by vinokurig** [comment]
Need to set 0.3.19 when Theia is released.

File: extensions/eclipse-che-theia-plugin-ext/src/common/che-protocol.ts [hunk 6]
```diff
@@ -54,6 +55,21 @@ export interface CheVariablesMain {
     $resolve(value: string): Promise<string | undefined>;
 }
 
+export interface CheTask {
+    registerTaskRunner(type: string, runner: che.TaskRunner): Promise<che.Disposable>;
+    fireTaskExited(id: number): Promise<void>;
+    $runTask(id: number, config: che.TaskConfiguration, ctx?: string): Promise<void>;
+    $killTask(id: number): Promise<void>;
+    $killTask(id: number): Promise<void>;
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply extensions/eclipse-che-theia-plugin-ext/src/common/che-protocol.ts]: fixed

File: extensions/eclipse-che-theia-plugin/src/che-proposed.d.ts [hunk 7]
```diff
@@ -87,4 +87,42 @@ declare module '@eclipse-che/plugin' {
...
+        readonly config: TaskConfiguration
+    }
+
+    export interface TaskConfiguration {
+        readonly type: string;
+        /** A label that uniquely identifies a task configuration */
+        readonly label: string;
+
+        /** Additional task type specific properties. */
+        readonly [key: string]: string;
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply extensions/eclipse-che-theia-plugin/src/che-proposed.d.ts]: fixed

**Pull request merged**